packages/contracts/contracts/deploy/PayableDAppFactory.sol

```diff
@@ -0,0 +1,132 @@
+// SPDX-License-Identifier: Apache-2.0
+/// @title ERC20 DApp Factory
+pragma solidity ^0.8.13;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IConsensus} from "@cartesi/rollups/contracts/consensus/IConsensus.sol";
+import {CartesiDApp} from "@cartesi/rollups/contracts/dapp/CartesiDApp.sol";
+import {ICartesiDApp} from "@cartesi/rollups/contracts/dapp/ICartesiDApp.sol";
+
+import {ICartesiIPFSDAppFactory} from "../factory/ICartesiIPFSDAppFactory.sol";
+import {IFinancialProtocol} from "../protocol/IFinancialProtocol.sol";
+import {IMachineProtocol} from "../protocol/IMachineProtocol.sol";
+import {IPayableDAppFactory} from "./IPayableDAppFactory.sol";
+
+/// @notice This contract implements an IPayableDAppFactory based on a IERC20 token
+contract PayableDAppFactory is
+    IPayableDAppFactory,
+    IFinancialProtocol,
+    IMachineProtocol,
+    Ownable
+{
+    ICartesiIPFSDAppFactory public immutable factory;
+    IERC20 public immutable token;
+    IConsensus public immutable consensus;
+    uint256 public immutable price;
+
+    mapping(ICartesiDApp => uint256) public runway;
+
+    /// @notice Construct a new ERC20DAppFactory with the given token as payment method
+    /// @param _factory The factory to delegate calls to instantiate the dapp
+    /// @param _consensus The consensus contract to use
+    /// @param _price The price per second of execution of the dapp node
+    constructor(
+        address _owner,
+        ICartesiIPFSDAppFactory _factory,
+        IERC20 _token,
+        IConsensus _consensus,
+        uint256 _price
+    ) {
+        factory = _factory;
+        token = _token;
+        consensus = _consensus;
+        price = _price;
+
+        // transfer ownership to the give owner
+        transferOwnership(_owner);
+    }
+
+    /// @dev This function deploys a new application, and registers it with the payment system
+    function newApplication(
+        address _dappOwner,
+        bytes32 _templateHash,
+        string memory _cid,
+        uint256 _runway
+    ) external override returns (CartesiDApp) {
+        // delegate call to factory, owner must be this contract, so we can register with payment system
+        CartesiDApp dapp = factory.newApplication(
+            consensus,
+            address(this),
+            _templateHash,
+            _cid
+        );
+
+        // now we transfer ownership to the given owner
+        dapp.transferOwnership(_dappOwner);
+
+        // set initial runway
+        this.extendRunway(dapp, _runway);
+
+        // return application
+        return dapp;
+    }
+
+    function importApplication(
```

Comment:
I think this should only be callable by the DApp owner! Or it should check if this application was already important. I'm worried about anyone being able to call this repeatedly and change the machine location.

(though `extendRunway` should be callable by anyone)

Reply:
True. Beware that the CID if content addressable, so if it's a different CID it's a different machine, and consequently a different machine hash. So when the validator try to start the node it will fail because it won't match the onchain hash.